sirepo/package_data/template/controls/parameters.py.jinja

```diff
@@ -127,7 +132,17 @@ def run_simulation(correctors):
         for k in values:
             lattice = re.sub('{' + k + '}', str(values[k]), lattice)
         f.write(lattice)
-    return os.system('madx in.madx > madx.log 2> madx.err') == 0
+   
+    success = os.system('madx in.madx > madx.log 2> madx.err') == 0
+    if not success:
+        return False
+
+    cmd = 'grep "is lost" madx.log  | wc'
+    res = subprocess.check_output(cmd, shell=True)
```

Comment:
Here is place where subprocess() is called. Mad-x doesn't fail if all particles are lost, but the data files will be missing and the simulation should be marked as an optimization failure. The only indicator we could find was to sum up the "is lost" lines in the madx.log file and compare against the original particle count. There is also an os.system() call above to invoke madx. @robnagler would there be a better way to do these call-outs?

Reply:
I'd recommend something like this, but definitely check the conditionals: 

```py
with open('{}') as f:
    c = 0
    for l in f.readline():
        if 'is lost' in l:
            c += 1
            if c > {{ particleCount }}:
                return False
    return True
```

I don't necessarily trust statements like "lost == particleCount". [Here's why.](https://github.com/radiasoft/pykern/wiki/DesignHints#ensure-loops-exit)

If you use wc, you can pass `-l`, if you want a count, but you could also pass `-c` to grep. However, as you see, you don't need a subprocess at all if it is simple parsing. This is done in shell scripts, but Python has reasonable enough parsing to make the code simpler and more readable to just use open/read.

I also am concerned about os.system. It's a bad example. It's always better to avoid shell parsing. There have been numerous security issues related to this. Rather, use something parsed:

```py
p = subprocess.run(('madx', 'in.madx'), text=True)
if p.returncode != 0:
    return False
```

Note that you avoid the open() and read() because you have p.stdout to decode and parse.

Reply:
```python
p = subprocess.run(('madx', 'in.madx'), stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
pkio.write_text('madx.log', p.stdout)
pkio.write_text('madx.err', p.stderr)
if p.returncode != 0:
    return False

lost = pkio.read_text('madx.log').count('is lost')
return lost != {{ particleCount }}
```
How about ^
Seems to work well